airflow/jobs.py

```diff
@@ -932,13 +929,19 @@ def _execute_task_instances(self,
         :type states: Tuple[State]
         :return: None
         """
-        # Get all the relevant task instances
+        # Get all the queued task instances from associated with scheduled
+        # DagRuns.
         TI = models.TaskInstance
         task_instances_to_examine = (
             session
             .query(TI)
             .filter(TI.dag_id.in_(simple_dag_bag.dag_ids))
             .filter(TI.state.in_(states))
+            .join(DagRun, and_(TI.dag_id == DagRun.dag_id,
+                               TI.execution_date == DagRun.execution_date,
+                               DagRun.state == State.RUNNING,
+                               DagRun.run_id.like(DagRun.ID_PREFIX + '%')))
```

Comment:
Yeah, this is required to prevent scheduling of tasks up for retry in failed DAG runs. It does require a full table scan since it lacks indexes. It hasn't been an issue as the DAG run table is relatively small, but it'll be a problem later on. Let's add indexes for that table then?